File: Python/example_18/Lotka_Volterra.py
import numpy as np
from scipy.integrate import odeint


############################
### Case study 18
############################
def Lotka_Volterra(x):
    # ####################################################################### #
    # LOTKA_VOLTERRA Analytic solution predator-prey populations in food web  #
    # Please check the work of Massoud et al. in Ecology Letters, 2017        #
    # ####################################################################### #

    if not hasattr(Lotka_Volterra, "initialized"):                                  # Store local variables in memory
        Lotka_Volterra.y0 = [30, 4]                                                 # Initial conditions: initial population of prey and predator
        Lotka_Volterra.tout = np.arange(0, 20 + 1/12, 1/12)                         # Time vector: from 0 to 20 years with monthly intervals
        def dydt_func(y, t, alpha, beta, gamma_, delta):                            # Define the Lotka-Volterra equations as a function
            dydt1 = alpha * y[0] - beta * y[0] * y[1]                               # y[0] = prey population, y[1] = predator population
            dydt2 = -gamma_ * y[1] + delta * y[0] * y[1]
            return [dydt1, dydt2]                                                   # Return dydt of the two states

        Lotka_Volterra.dydt = dydt_func                                             # Store the dydt function
        Lotka_Volterra.initialized = True                                           # Flag to indicate that initialization is complete        

    y = Lotka_Volterra.y0                                                           # Unpack the parameters (alpha, beta, gamma_, delta) and add initial states
    result = odeint(Lotka_Volterra.dydt, y, Lotka_Volterra.tout, args = tuple(x))   # Solve the system of differential equations using odeint
    Y = result[1:, :2]                                                              # Return only the prey and predator populations
    Y_vector = Y.T.reshape(-1,1)                                                    # Flatten to return as a one-dimensional vector

    return Y_vector.flatten(), Y_vector.flatten()

File: MATLAB/example_18/lotka_volterra.m
function [ Y ] = lotka_volterra ( x )
% Lotka-Volterra model of population dynamics

persistent dydt y0 tout;    % Retain variables in memory after first call

if isempty(dydt)
    %% Initial conditions (initial population of prey and predator population)
    y0 = [30 4];
    %% Print time in years (start at t = 0 )
    tout = [ 0 : 1/12 : 20 ];
    % Define dydt --> the Lotka-Volterra equations differential equations
    % dydt = inline('[ alpha * y(1) - beta * y(1) * y(2) ; -gamma * y(2) + delta * y(1) * y(2) ]', ...
    %   't', 'y','alpha','beta','gamma','delta');
    % We apply a trick so we can use built-in ODE solver with inline function and additional parameters
    % We pass parameters as state variables (three to six) and specify
    % their partial derivatives as zero so their values do not change
%    dydt = inline('[ y(3)*y(1)-y(4)*y(1)*y(2) ; -y(5)*y(2)+y(6)*y(1)*y(2) ; zeros(4,1) ]', 't', 'y');
    dydt = @(t,y) [ y(3)*y(1)-y(4)*y(1)*y(2) ; -y(5)*y(2)+y(6)*y(1)*y(2) ; zeros(4,1) ];
end

%% Unpack the parameters of x in order of alpha, beta, gamma and delta and add to initial states
y = [ y0 x ]';
%% Now use ode45 time-variable integration method to solve system of two equations (monthly print step)
[ t , Y ] = ode45 ( dydt , tout , y ); 
% Remove the first row of Y (initial state at t = 0) and return output as one vector
Y = Y(2:end,1:2); Y = Y(:);